src/file_pipe_log/pipe.rs

```diff
@@ -155,12 +156,22 @@ impl<F: FileSystem> SinglePipe<F> {
         Ok(())
     }
 
-    fn new_file(&self, seq: FileSeq) -> Result<(PathId, F::Handle)> {
+    /// Recycles one obsolete file from the recycled file list.
+    fn recycle_file(&self) -> (usize, Option<File<F>>) {
+        let mut recycled_files = self.recycled_files.write();
+        let f = recycled_files.pop_front();
+        (recycled_files.len(), f)
+    }
+
+    /// Creates a new log file according to the given [`FileSeq`].
+    ///
+    /// If `recycle_file` is valid, it will reuse it as priority.
+    fn new_file(&self, seq: FileSeq, recycle_file: Option<File<F>>) -> Result<(PathId, F::Handle)> {
         let new_file_id = FileId {
             seq,
             queue: self.queue,
         };
-        if let Some(f) = self.recycled_files.write().pop_front() {
+        if let Some(f) = recycle_file {
             let src_path = self.paths[f.path_id].join(build_recycled_file_name(f.seq));
```

Comment:
Better move these parts into `recycle_file` too, and update the counters in `recycle_file`. The reason is we try to put the metrics update the same level as the file collection update.